File: LeetCode/Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    pair<bool,int> dfs(TreeNode* root) {
        if (!root) return {true, 0};

        auto left = dfs(root->left);
        auto right = dfs(root->right);

        bool balanced = abs(left.second - right.second) <= 1;

        return {balanced & left.first & right.first, max(left.second, right.second) + 1};
    }
public:
    bool isBalanced(TreeNode* root) {
        return dfs(root).first;
    }
};

File: LeetCode/Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def get_balanced_height(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        left_height = self.get_balanced_height(root.left)
        right_height = self.get_balanced_height(root.right)

        if left_height == -1 or right_height == -1:
            return -1

        if abs(left_height - right_height) > 1:
            return -1

        return max(left_height, right_height) + 1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if self.get_balanced_height(root) == -1:
            return False
        else:
            return True
        